runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -336,11 +356,29 @@ public void clear(K key, W window) {
             BagState<V> bagState =
                 stateInternals.state(namespace, StateTags.bag(userStateId, valueCoder));
             bagState.clear();
+
+            return generateAndRegisterCacheKey();
           } finally {
             stateBackendLock.unlock();
           }
         }
 
+        @Override
+        public Iterable<ByteString> getCacheTokens() {
+          return cacheTokens.asMap().values();
+        }
+
+        @Override
+        public void clearCacheTokens() {
+          cacheTokens.invalidateAll();
+        }
+
+        private ByteString generateAndRegisterCacheKey() {
+          ByteString cacheToken = ByteString.copyFrom(cacheTokenGenerator.getId(), Charsets.UTF_8);
+          cacheTokens.put(userStateId, cacheToken);
```

Comment:
Correct me if I am wrong there. The bag states are also separated by window, would it make sense to include window in the cache key?

Reply:
The same cache token can be handed out for all state cells across all windows. On the SDK side we need to scope the cache per window because state needs to be retrieved at least once per window.

On this note, I will also remove the cache token per user state because we just need one cache token for the key range (i.e. operator instance) for all user state. The reason is that user state does not get invalidated for other reasons than an updated key range. For side input, each side input needs a separate cache token to be able to invalidate them individually in case they get updated.